src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -237,6 +238,38 @@ func validateRecaptcha(
 	return nil
 }
 
+// generates username for registration if none provided.
+// Sequentially checks for available username and returns the lowest available one.
+func generateUsername(
```

Comment:
So I believe this is analogous to how Synapse does username generation, but it seems inherently slow.

I'd need someone from synapse dev to weigh in on whether its for some reason a bit idea to do so, but my suggestion for username generation would instead be:

* Check total number of users in DB on startup. Save in a variable.
* When registering a user, increment this variable.
* When generating a username, set the localpart to this number, then increment.

This would be much more efficient, and I'm interested in hearing anyone's comments on why this wouldn't work.

Reply:
@anoadragon453 
Lets say database is empty. I am the first user and registers with a username as 2 by explicitly specifying. Then we get a second a user who registers without username now by your logic he should be given a localpart of 2 since he is 2nd user but it fails. 2 users cant have same localpart.

We can give something really strange like a timestamp for username(after a bit of regExp filtering) and then let register function validate it. If user is found generate a new timestamp(it wd be different since time passed while checking). The probability that anyone would give such username explicitly decreases exponentially. At max we have to do it 2 times. 99% cases one will do.

Reply:
@anoadragon453 Something like that can be very efficiently done by goroutines . But that may require passing another parameter from the setup of clientapi till register function.
This is the reason why I feel it would seem out of place. Thoughts ?